### PR TITLE
HMAN-29 -Build generic Inbound Topic Subscriber

### PR DESCRIPTION
**Pull request by padmapriyanalam**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/HMAN-29

### Change description ###
Build generic Inbound Topic Subscriber

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/hmc/ApplicationParams.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+import lombok.Getter;
+import org.springframework.beans.factory.annotation.Value;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+@Getter
+@Named
+@Singleton
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/ApplicationParams.java:11]:
```suggestion
@Component
```
Then we can delete the `javax.inject` dependency from the `build.gradle`

File: build.gradle [hunk 2]
```diff
@@ -285,12 +286,19 @@ dependencies {
   implementation group: 'org.projectlombok', name: 'lombok', version: versions.lombok
   annotationProcessor group: 'org.projectlombok', name: 'lombok', version: versions.lombok
 
+  implementation group: 'com.github.tomakehurst', name: 'wiremock', version: '2.27.2'
```
Comment (danlysiak) [inline build.gradle:289]: I doubt WireMock is being used in implementation? Can we move this to the appropriate scope please

File: charts/hmc-cft-hearing-service/values.yaml [hunk 3]
```diff
@@ -8,3 +8,16 @@ java:
       secrets:
         - name: AppInsightsInstrumentationKey
           alias: azure.application-insights.instrumentation-key
+        - name: hmc-servicebus-connection-string
+          alias: HMC_SERVICE_BUS_CONNECTION_STRING
+        - name: hmc-servicebus-subscription
+          alias: HMC_SERVICE_BUS_SUBSCRIPTION
+        - name: hmc_servicebus-topic
```
Comment (danlysiak) [inline charts/hmc-cft-hearing-service/values.yaml:15]:
```suggestion
        - name: hmc-servicebus-topic
```

File: src/main/java/uk/gov/hmcts/reform/hmc/config/MessageReaderFromTopicConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        log.info("Starting the processor");
+        processorClient.start();
+
+        try {
+            TimeUnit.SECONDS.sleep(10);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        log.info("Stopping and closing the processor");
+        processorClient.close();
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/config/MessageReaderFromTopicConfiguration.java:55]: We don't want to be waiting and/or stopping the processing - as long as the CFT Hearing Service is running it should be accepting and processing messages

File: src/main/java/uk/gov/hmcts/reform/hmc/config/MessageReaderFromTopicConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    private static void processMessage(ServiceBusReceivedMessageContext context) {
+        ServiceBusReceivedMessage message = context.getMessage();
+        log.info("Received message with id '{}', message: {}", message.getMessageId(), message.getBody());
+    }
+
+    private static void processError(ServiceBusErrorContext context, CountDownLatch countdownLatch) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/config/MessageReaderFromTopicConfiguration.java:63]: Fine for now until we have any specific error handling reqs 👍

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -42,3 +42,9 @@ spring:
 azure:
   application-insights:
     instrumentation-key: ${hmc.AppInsightsInstrumentationKey:00000000-0000-0000-0000-000000000000}
+
+jms:
+  servicebus:
+    connection-string: ${HMC_SERVICE_BUS_CONNECTION_STRING:Endpoint=sb://hmc-servicebus-demo.servicebus.windows.net/;SharedAccessKeyName=[KEYNAME];SharedAccessKey=[KEY]}
```
Comment (danlysiak) [inline src/main/resources/application.yaml:48]:
```suggestion
    connection-string: ${HMC_SERVICE_BUS_CONNECTION_STRING}
```

File: src/main/resources/application.yaml [hunk 7]
```diff
@@ -42,3 +42,9 @@ spring:
...
+    topic-name: ${HMC_SERVICE_BUS_TOPIC:hmc-to-cft-demo}
+    subscription-name: ${HMC_SERVICE_BUS_SUBSCRIPTION:hmc-subs-to-cft-demo}
```
Comment (danlysiak) [inline src/main/resources/application.yaml:50]:
```suggestion
    connection-string: ${HMC_SERVICE_BUS_CONNECTION_STRING}
    topic-name: ${HMC_SERVICE_BUS_TOPIC}
    subscription-name: ${HMC_SERVICE_BUS_SUBSCRIPTION}
```
Only need these changes on local please 🙏

File: src/test/resources/application-itest.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+azure:
+  jms:
+    servicebus:
+      connection-string: ${HMC_SERVICE_BUS_CONNECTION_STRING:Endpoint=sb://destination1.servicebus.windows.net;SharedAccessKeyName=[KEYNAME];SharedAccessKey=[KEY]}
+      topic-name: ${HMC_SERVICE_BUS_TOPIC:hmc-to-cft}
+      subscription-name: {HMC_SERVICE_BUS_SUBSCRIPTION}
```
Comment (danlysiak) [inline src/test/resources/application-itest.yaml:6]:
```suggestion
      subscription-name: ${HMC_SERVICE_BUS_SUBSCRIPTION:hmc-subs-to-cft}
```

Comment (padmapriyanalam) [reply src/test/resources/application-itest.yaml:6]: added

File: src/main/resources/application.yaml (hunk 7 above)
Comment (padmapriyanalam) [reply src/main/resources/application.yaml:50]: changed

File: charts/hmc-cft-hearing-service/values.yaml (hunk 3 above)
Comment (padmapriyanalam) [reply charts/hmc-cft-hearing-service/values.yaml:15]: updated

File: build.gradle (hunk 2 above)
Comment (padmapriyanalam) [reply build.gradle:289]: removed

File: src/main/java/uk/gov/hmcts/reform/hmc/config/MessageReaderFromTopicConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    public void readMessageFromTopic() {
+        CountDownLatch countdownLatch = new CountDownLatch(1);
+
+        // Create an instance of the processor through the ServiceBusClientBuilder
+        ServiceBusProcessorClient processorClient = new ServiceBusClientBuilder()
+            .connectionString(applicationParams.getConnectionString())
+            .processor()
+            .topicName(applicationParams.getTopicName())
+            .subscriptionName(applicationParams.getSubscriptionName())
+            .processMessage(MessageReaderFromTopicConfiguration::processMessage)
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/config/MessageReaderFromTopicConfiguration.java:42]:
```suggestion
            .processMessage(this::processMessage)
```

Sorry didn't notice on first run-through that `processMessage()` and `processError()` are both marked as `static`. Can we remove the `static` from them please - updating this line should allow that.

File: src/main/resources/application.yaml (hunk 7 above)
Comment (danlysiak) [reply src/main/resources/application.yaml:50]: Re-opening for visibility as think this commit hasn't been pushed yet?

**Review by danlysiak — APPROVED** [review]

**Review by noronhaa — APPROVED** [review]

**Pull request merged**